File: app/src/main/java/com/tanacom/introrealm/java/MainActivity.java
//package com.tanacom.introrealm.java;
//
//import android.os.Bundle;
//import android.widget.Toast;
//
//import androidx.annotation.NonNull;
//import androidx.annotation.Nullable;
//import androidx.appcompat.app.AppCompatActivity;
//import androidx.fragment.app.DialogFragment;
//import androidx.recyclerview.widget.DividerItemDecoration;
//import androidx.recyclerview.widget.ItemTouchHelper;
//import androidx.recyclerview.widget.LinearLayoutManager;
//import androidx.recyclerview.widget.RecyclerView;
//
//import com.google.android.material.floatingactionbutton.FloatingActionButton;
//import com.tanacom.introrealm.R;
//
//import io.realm.Realm;
//
//public class MainActivity extends AppCompatActivity implements CountryListener {
//
//
//    private Realm realm;
//    private RecyclerView recyclerView;
//    private FloatingActionButton fab;
//
//    @Override
//    protected void onCreate(@Nullable Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//
//        setContentView(R.layout.activity_main);
//
//        init();
//    }
//
//    private void init() {
//
//        realm = Realm.getDefaultInstance();
//
//        recyclerView = findViewById(R.id.rv);
//
//        DialogFragment taskDialog = new CountryDialog();
//
//        fab = findViewById(R.id.fab);
//
//        setRecyclerView();
//
//        fab.setOnClickListener(
//                view -> {
//                    taskDialog.show(getSupportFragmentManager(), "New Country");
//
//                }
//        );
//
//
//    }
//
//    @Override
//    public void onAddCountry(String country) {
////
////        SecureRandom rand = new SecureRandom();
////        // from 0 to 99999
////        int countryID = rand.nextInt(100000);
////
////        CountryDataHelper.addCountry(realm, countryID, country);
//
//        CountryDataHelper.addCountry(realm, country);
//
//    }
//
//    private void setRecyclerView() {
//
//        CountryAdapter adapter = new CountryAdapter(realm.where(CountryModel.class).findAll());
//
//        recyclerView.setAdapter(adapter);
//
//        recyclerView.setLayoutManager(new LinearLayoutManager(this));
//
//        recyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.HORIZONTAL));
//
//        TouchHelperCallback th = new TouchHelperCallback();
//
//        ItemTouchHelper ith = new ItemTouchHelper(th);
//
//        ith.attachToRecyclerView(recyclerView);
//
//    }
//
//    @Override
//    public void onCancel(DialogFragment dialogFragment) {
//
//        Toast.makeText(this, "Cancelled", Toast.LENGTH_SHORT).show();
//
//    }
//
//    private class TouchHelperCallback extends ItemTouchHelper.SimpleCallback {
//
//        public TouchHelperCallback() {
//
//            super(ItemTouchHelper.UP | ItemTouchHelper.DOWN, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT);
//
//        }
//
//        @Override
//        public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
//            return false;
//        }
//
//        @Override
//        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
//
//            CountryDataHelper.deleteCountry(realm, viewHolder.getItemId());
//
//        }
//
//    }
//
//
//}

File: app/src/main/java/com/tanacom/introrealm/kotlin/MainActivity.kt
package com.tanacom.introrealm.kotlin

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.android.volley.AuthFailureError
import com.android.volley.Response
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.tanacom.introrealm.R
import com.tanacom.introrealm.kotlin.Constants.ACCEPT
import com.tanacom.introrealm.kotlin.Constants.AUTHORIZATION
import com.tanacom.introrealm.kotlin.Constants.ENDPOINTS
import com.tanacom.introrealm.kotlin.Constants.TAGG
import com.tanacom.introrealm.kotlin.CountryDataHelper.addCountry
import com.tanacom.introrealm.kotlin.CountryDataHelper.deleteCountry
import com.tanacom.introrealm.kotlin.CountryDataHelper.syncCountry
import io.realm.Realm
import org.json.JSONArray
import org.json.JSONObject


class MainActivity : AppCompatActivity(), CountryListener {
    private var realm: Realm? = null
    private var recyclerView: RecyclerView? = null
    private var fab: FloatingActionButton? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        init()
    }

    private fun init() {
        realm = Realm.getDefaultInstance()
        recyclerView = findViewById(R.id.rv)
        val taskDialog: DialogFragment = CountryDialog()
        fab = findViewById(R.id.fab)
        setRecyclerView()

        fab!!.setOnClickListener { view: View? ->
            taskDialog.show(
                supportFragmentManager, TAGG
            )
        }


    }

    private fun setRecyclerView() {
        val adapter = CountryAdapter(
            realm!!.where(
                CountryModel::class.java
            ).findAll()
        )
        recyclerView!!.adapter = adapter
        recyclerView!!.layoutManager = LinearLayoutManager(this)
        recyclerView!!.addItemDecoration(
            DividerItemDecoration(
                this,
                DividerItemDecoration.VERTICAL
            )
        )

        recyclerView!!.adapter!!.notifyDataSetChanged()

        val th = TouchHelperCallback()
        val ith = ItemTouchHelper(th)
        ith.attachToRecyclerView(recyclerView)
    }

    override fun onAddCountry(country: String?) {
        addCountry(realm!!, country)

    }

    override fun onCancel(dialogFragment: DialogFragment?) {
        Toast.makeText(this, "Cancelled", Toast.LENGTH_SHORT).show()
    }

    override fun onSync() {
        getCountries()

    }

    private fun getCountries() {
        val queue = Volley.newRequestQueue(this)

        val url = ENDPOINTS

        val getRequest: StringRequest = object : StringRequest(
            Method.GET, url,
            Response.Listener<String?> { response -> // response
                Log.d("Response", response)

                handleResponse(response)
            },

            Response.ErrorListener { error ->
                Log.d("ERROR", "error => $error")
                Toast.makeText(
                    applicationContext,
                    "Try Again",
                    Toast.LENGTH_SHORT
                ).show()
            }
        ) {
            @Throws(AuthFailureError::class)
            override fun getHeaders(): Map<String, String> {
                val params: MutableMap<String, String> = HashMap()
                params["Authorization"] = AUTHORIZATION
                params["Accept"] = ACCEPT
                return params
            }
        }
        queue.add(getRequest)
    }

    private fun handleResponse(response: String) {
        val jsonArray = JSONArray(response)
        for (i in 0 until jsonArray.length()) {
            val jsonObject: JSONObject = jsonArray[i] as JSONObject

            syncCountry(realm!!, jsonObject.getString("country_name"))

        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val inflater = menuInflater

        inflater.inflate(R.menu.menu, menu)

        return true

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.sync -> {
                getCountries()
                true

            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    private inner class TouchHelperCallback : ItemTouchHelper.SimpleCallback(
        ItemTouchHelper.UP or ItemTouchHelper.DOWN,
        ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
    ) {
        override fun onMove(
            recyclerView: RecyclerView,
            viewHolder: RecyclerView.ViewHolder,
            target: RecyclerView.ViewHolder
        ): Boolean {
            return false
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
            deleteCountry(realm!!, viewHolder.itemId)
        }
    }


}